demos/bluetooth/mesh/thingy52/thingy52_srv/src/model_handler.c

```diff
@@ -0,0 +1,348 @@
+
+static struct k_delayed_work led_fade_work;
+
+static void led_fade_work_handler(struct k_work *work)
+{
+	static bool do_fade;
+
+	if (!cur_rgb_msg.color.red && !cur_rgb_msg.color.green &&
+	    !cur_rgb_msg.color.blue) {
+		do_fade = false;
+		printk("LED fade complete\n");
+	} else if (cur_rgb_msg.delay == 0xFFFF) {
+		cur_rgb_msg.color.red = (cur_rgb_msg.color.red * 90) / 100;
+		cur_rgb_msg.color.green = (cur_rgb_msg.color.green * 90) / 100;
+		cur_rgb_msg.color.blue = (cur_rgb_msg.color.blue * 90) / 100;
+
+		sx1509b_pwm_pin_set(dev.io_expander, RED_LED,
+				    cur_rgb_msg.color.red);
+		sx1509b_pwm_pin_set(dev.io_expander, GREEN_LED,
+				    cur_rgb_msg.color.green);
+		sx1509b_pwm_pin_set(dev.io_expander, BLUE_LED,
+				    cur_rgb_msg.color.blue);
+
+		k_delayed_work_submit(&led_fade_work, K_MSEC(20));
+	} else if (cur_rgb_msg.delay > 0) {
+		if (cur_rgb_msg.delay <= 300 || do_fade) {
+			cur_rgb_msg.color.red =
+				(cur_rgb_msg.color.red * 80) / 100;
+			cur_rgb_msg.color.green =
+				(cur_rgb_msg.color.green * 80) / 100;
+			cur_rgb_msg.color.blue =
+				(cur_rgb_msg.color.blue * 80) / 100;
+
+			sx1509b_pwm_pin_set(dev.io_expander, RED_LED,
+					    cur_rgb_msg.color.red);
+			sx1509b_pwm_pin_set(dev.io_expander, GREEN_LED,
+					    cur_rgb_msg.color.green);
+			sx1509b_pwm_pin_set(dev.io_expander, BLUE_LED,
+					    cur_rgb_msg.color.blue);
+
+			k_delayed_work_submit(&led_fade_work, K_MSEC(20));
+		} else {
+			do_fade = true;
+			k_delayed_work_submit(&led_fade_work,
+					      K_MSEC(cur_rgb_msg.delay - 300));
+		}
+	}
+}
+
+static void set_rgb_led(struct bt_mesh_thingy52_rgb_msg rgb_msg)
+{
+	sx1509b_pwm_pin_set(dev.io_expander, RED_LED, rgb_msg.color.red);
+	sx1509b_pwm_pin_set(dev.io_expander, GREEN_LED, rgb_msg.color.green);
+	sx1509b_pwm_pin_set(dev.io_expander, BLUE_LED, rgb_msg.color.blue);
+
+	cur_rgb_msg = rgb_msg;
+
+	k_delayed_work_submit(&led_fade_work, K_NO_WAIT);
+}
+
+static void led_init(void)
+{
+	int err = 0;
+
+	for (int i = GREEN_LED; i <= RED_LED; i++) {
+		err |= sx1509b_pwm_pin_configure(dev.io_expander, i);
+	}
+
+	k_delayed_work_init(&led_fade_work, led_fade_work_handler);
+
+	if (err) {
+		printk("Initializing LEDs failed\n");
+	}
+}
+
+static void speaker_init(uint32_t pwm_frequency)
+{
+	int err = 0;
+	uint32_t pwm_period = 1000000U / pwm_frequency;
+
+	err |= gpio_pin_configure(dev.gpio_0, SPKR_PWR,
+				  GPIO_OUTPUT | GPIO_PULL_UP);
+	err |= pwm_pin_set_usec(dev.spkr_pwm, SPKR_PWM, pwm_period,
+				pwm_period / 2U, 0);
+
+	if (err) {
+		printk("Initializing speaker failed\n");
+	}
+}
+
+static bool next_rgb_work_idx_get(uint8_t *idx)
+{
+	*idx = 0;
+	uint16_t top_delay = 0xFFFF;
+	bool active_work_present = false;
+
+	for (int i = 0; i < ARRAY_SIZE(rgb_work); ++i) {
+		int32_t rem = k_delayed_work_remaining_ticks(&rgb_work[i].work);
+
+		if ((rem != 0) && (rgb_work[i].rgb_msg.delay < top_delay)) {
+			*idx = i;
+			top_delay = rgb_work[i].rgb_msg.delay;
+			active_work_present = true;
+		}
```

Comment:
I'm having some trouble getting what this rgb_work array is about. Is this scheduling colors based on their initial delay, with the color with the shortest delay going first? Why is the delay time the deciding factor, and not the remaining time or time of arrival? What happens in this scenario?
T=1 RX(red, delay=5)
T=2 RX(blue, delay=6)
T=3 RX(green, delay=2)
Do the colors go like this? 
T1. red
T2. red
T3. green
T4. green
T5. red
T6. blue
T7. blue
T8. none
?

Reply:
I can see that it might be a little bit fuzzy. The idea is that the color associated with the shortest delay should be the one occupying the large lightwell when there are multiple messages active in a node at once. 

In the scenario you have described the correct behaviour as far as I can see it.
 
I can take a look at the possibility  of using rem-time instead of the initial delay. I agree that it makes more sense.